File: questPresentation/src/main/java/ru/nekit/android/qls/window/AnswerWindow.java
package ru.nekit.android.qls.window;

import android.content.Context;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StyleRes;
import android.view.ViewGroup;

import ru.nekit.android.qls.R;
import ru.nekit.android.qls.quest.QuestContext;
import ru.nekit.android.qls.utils.ViewHolder;
import ru.nekit.android.qls.window.common.QuestWindow;

import static android.view.ViewGroup.LayoutParams.MATCH_PARENT;

public class AnswerWindow extends QuestWindow {

    private AnswerWindow(@NonNull QuestContext questContext, @NonNull String name,
                         @NonNull WindowContentViewHolder contentViewHolder,
                         @StyleRes int styleResId) {
        super(questContext, name, contentViewHolder, styleResId);
    }

    public enum Type {

        RIGHT,
        WRONG;

        public String getName() {
            return name();
        }
    }

    public static class Builder {

        private final QuestContext mQuestContext;
        @NonNull
        private final Type mType;
        @NonNull
        private AnswerWindow mWindow;
        @Nullable
        private AnswerWindowContainer mWindowContainer;
        @Nullable
        private ViewHolder mContentHolder, mToolContentHolder;
        @StyleRes
        private int mStyleResId = -1;

        public Builder(@NonNull QuestContext context, @NonNull Type type) {
            mQuestContext = context;
            mType = type;
        }

        @NonNull
        public Builder setContent(@NonNull ViewHolder content) {
            mContentHolder = content;
            return this;
        }

        @NonNull
        public Builder setContent(@LayoutRes int contentResId) {
            if (contentResId > 0) {
                mContentHolder = new ViewHolder(mQuestContext, contentResId);
            }
            return this;
        }

        @NonNull
        public Builder setToolContent(@NonNull ViewHolder content) {
            mToolContentHolder = content;
            return this;
        }

        @NonNull
        public Builder setToolContent(@LayoutRes int toolContentResId) {
            mToolContentHolder = new ViewHolder(mQuestContext, toolContentResId);
            return this;
        }

        @NonNull
        public Builder setStyle(@StyleRes int styleResId) {
            mStyleResId = styleResId;
            return this;
        }

        @NonNull
        public AnswerWindow create() {
            mWindowContainer = new AnswerWindowContainer(mQuestContext);
            if (mContentHolder != null) {
                mWindowContainer.contentContainer.addView(mContentHolder.view, MATCH_PARENT,
                        MATCH_PARENT);
            }
            if (mToolContentHolder != null) {
                mWindowContainer.toolContainer.addView(mToolContentHolder.view);
            }
            mWindow = new AnswerWindow(mQuestContext, mType.getName(), mWindowContainer, mStyleResId == -1 ? mType == Type.RIGHT ?
                    R.style.Window_RightAnswer : R.style.Window_WrongAnswer : mStyleResId);
            return mWindow;
        }
    }

    public static class AnswerWindowContainer extends WindowContentViewHolder {

        @NonNull
        final ViewGroup contentContainer, toolContainer;

        AnswerWindowContainer(@NonNull Context context) {
            super(context, R.layout.wc_answer);
            toolContainer = (ViewGroup) view.findViewById(R.id.container_tool);
            contentContainer = (ViewGroup) view.findViewById(R.id.container_content);
        }

        @Override
        protected int getCloseButtonId() {
            return R.id.btn_ok;
        }

    }
}

File: questPresentation/src/main/java/ru/nekit/android/qls/window/AnswerWindow.kt
package ru.nekit.android.qls.window

import android.support.annotation.LayoutRes
import android.support.annotation.StyleRes
import android.view.View
import android.widget.TextView
import ru.nekit.android.qls.R
import ru.nekit.android.qls.data.representation.getRepresentation
import ru.nekit.android.qls.domain.model.AnswerType
import ru.nekit.android.qls.domain.model.RecordType.RIGHT_ANSWER_BEST_TIME_UPDATE_RECORD
import ru.nekit.android.qls.domain.model.RecordType.RIGHT_ANSWER_SERIES_LENGTH_UPDATE_RECORD
import ru.nekit.android.qls.quest.QuestContext
import ru.nekit.android.qls.window.AnswerWindowType.RIGHT
import ru.nekit.android.qls.window.AnswerWindowType.WRONG
import ru.nekit.android.qls.window.common.ContentWithToolQuestWindow
import ru.nekit.android.utils.ViewHolder
import java.text.SimpleDateFormat
import java.util.*

private object AnswerWindow {

    fun open(questContext: QuestContext,
             type: AnswerWindowType,
             @StyleRes styleResId: Int,
             @LayoutRes contentResId: Int,
             @LayoutRes toolContentResId: Int) {
        ContentWithToolQuestWindow.Builder(questContext, type.name)
                .setStyle(styleResId)
                .setContent(contentResId)
                .setToolContent(toolContentResId)
                .create()
                .open()
    }

    fun open(questContext: QuestContext,
             type: AnswerWindowType,
             @StyleRes styleResId: Int,
             content: ViewHolder,
             @LayoutRes toolContentResId: Int) {
        ContentWithToolQuestWindow.Builder(questContext, type.name)
                .setStyle(styleResId)
                .setContent(content)
                .setToolContent(toolContentResId)
                .create()
                .open()
    }

}

enum class AnswerWindowType {

    RIGHT,
    WRONG;

}

object WrongAnswerWindow {

    fun open(questContext: QuestContext,
             @StyleRes styleResId: Int,
             @LayoutRes contentResId: Int,
             @LayoutRes toolContentResId: Int) {
        AnswerWindow.open(questContext, WRONG, styleResId, contentResId, toolContentResId)
    }

    fun openSimple(questContext: QuestContext) {
        AnswerWindow.open(questContext, WRONG,
                R.style.Window_WrongAnswer_Simple,
                ViewHolder(questContext, R.layout.wc_wrong_answer_simple_content).apply {
                    buildTitle(questContext, view, AnswerType.WRONG)
                },
                R.layout.wc_wrong_answer_tool_simple_content)
    }

    fun open(questContext: QuestContext,
             @StyleRes styleResId: Int,
             content: ViewHolder,
             @LayoutRes toolContentResId: Int) {
        AnswerWindow.open(questContext, WRONG, styleResId, content, toolContentResId)
    }

    fun openDefault(questContext: QuestContext) {
        WrongAnswerWindow.open(
                questContext,
                R.style.Window_WrongAnswer,
                ViewHolder(questContext, R.layout.wc_wrong_answer_content).apply {
                    buildTitle(questContext, view, AnswerType.WRONG)
                },
                R.layout.wc_wrong_answer_tool_simple_content
        )
    }
}

private fun buildTitle(questContext: QuestContext, view: View, answerType: AnswerType) {
    (view.findViewById(R.id.tv_title) as TextView).apply {
        text = answerType.getRepresentation().getRandomString(questContext)
    }
}


object RightAnswerWindow {

    fun open(questContext: QuestContext,
             @StyleRes styleResId: Int,
             @LayoutRes contentResId: Int,
             @LayoutRes toolContentResId: Int) {
        AnswerWindow.open(questContext, RIGHT, styleResId, contentResId, toolContentResId)
    }

    fun openSimple(questContext: QuestContext) {
        AnswerWindow.open(questContext, RIGHT,
                R.style.Window_RightAnswer_Simple,
                ViewHolder(questContext, R.layout.wc_right_answer_simple_content).apply {
                    buildTitle(questContext, view, AnswerType.RIGHT)
                },
                R.layout.wc_right_answer_tool_simple_content)
    }

    fun open(questContext: QuestContext,
             @StyleRes styleResId: Int,
             content: ViewHolder,
             @LayoutRes toolContentResId: Int) {
        AnswerWindow.open(questContext, RIGHT, styleResId, content, toolContentResId)
    }

    fun openDefault(questContext: QuestContext) {
        questContext.apply {
            pupil { pupil ->
                questHistory { questHistory ->
                    questPreviousHistoryWithBestSessionTime(questHistory!!) { previousHistory ->
                        questStatisticsReport { report ->
                            RightAnswerWindow.open(
                                    this,
                                    R.style.Window_RightAnswer,
                                    ViewHolder(this, R.layout.wc_right_answer_content).apply {
                                        val dateFormat = SimpleDateFormat(
                                                getString(R.string.right_answer_timer_formatter),
                                                Locale.getDefault())
                                        ArrayList<String>().let { list ->
                                            list.add("${pupil.name}, ${
                                            AnswerType.RIGHT.getRepresentation().getRandomString(questContext).toLowerCase()
                                            }")
                                            list.add(String.format(questContext.getString(R.string.right_answer_time_formatter),
                                                    dateFormat.format(questHistory.sessionTime)))
                                            if (previousHistory != null && (questHistory.recordTypes and RIGHT_ANSWER_BEST_TIME_UPDATE_RECORD.value) != 0)
                                                list.add(String.format(questContext.getString(R.string.right_answer_best_time_update_formatter),
                                                        dateFormat.format(previousHistory.sessionTime)))
                                            if ((questHistory.recordTypes and RIGHT_ANSWER_SERIES_LENGTH_UPDATE_RECORD.value) != 0)
                                                list.add(String.format(getString(R.string.right_answer_series_length_update_formatter),
                                                        report.rightAnswerSeriesCount))
                                            if (questHistory.rewards.isNotEmpty()) {
                                                list.add("Награды: ${questHistory.rewards.size}")
                                                questHistory.rewards.forEach {
                                                    list.add(it.getRepresentation().getString(questContext))
                                                }
                                            }
                                            (view.findViewById(R.id.tv_title) as TextView).apply {
                                                text = list.joinToString("\n")
                                            }
                                        }
                                    },
                                    R.layout.wc_right_answer_tool_simple_content
                            )
                        }
                    }
                }
            }
        }
    }
}